### PR TITLE
changed: store rates in an array in SingleWellState

### PR DESCRIPTION
**Pull request by akva2**
introduce an enum for indexing into the array. this again allows us to coalesce 4 parallel reductions into one.

another micro optimization.

### CODE REVIEW

**Comment by akva2** [comment]
jenkins build this please

File: opm/simulators/wells/SingleWellState.hpp [hunk 1]
```diff
@@ -85,10 +82,15 @@ class SingleWellState {
     double bhp{0};
     double thp{0};
     double temperature{0};
-    double dissolved_gas_rate{0};
-    double dissolved_gas_rate_in_water{0};
-    double vaporized_oil_rate{0};
-    double vaporized_wat_rate{0};
+
+    std::array<double,4> rates{};
```
Comment (bska) [inline opm/simulators/wells/SingleWellState.hpp:86]: I'm not sure `rates` is the best name for this data member.  There are multiple other quantities in `SingleWellState` that could potentially be named `rates` as well.  As written, this is a more specialised collection that exclusively holds dissolved and vaporised flow rates and I think the name should reflect that more restricted purpose.  That said, I don't have a good suggestion for another name.  Something like `dissolved_rates` would be close, but might suggest that the vaporised components are excluded.  `Phase_mixing_rates` or just `mixing_rates` might be an alternative, but we've not used that notion anywhere else.

**Review by bska — COMMENTED** [review]
I think this is good save for some naming concerns.  I'm going to solicit opinions from @atgeirr, @GitPaean, and @totto82 on that matter.

File: opm/simulators/wells/MultisegmentWell_impl.hpp [hunk 2]
```diff
@@ -1545,10 +1545,7 @@ namespace Opm
         this->linSys_.clear();
 
         auto& ws = well_state.well(this->index_of_well_);
-        ws.dissolved_gas_rate = 0;
-        ws.dissolved_gas_rate_in_water = 0;
-        ws.vaporized_oil_rate = 0;
-        ws.vaporized_wat_rate = 0;
+        ws.rates = {};
```
[the remark this answers is not in the document]
Comment (akva2) [reply opm/simulators/wells/MultisegmentWell_impl.hpp]: done.

File: opm/simulators/wells/SingleWellState.hpp (hunk 1 above)
Comment (GitPaean) [reply opm/simulators/wells/SingleWellState.hpp:86]: This is also my only concern. Maybe `phase_mixing_rates`  will be the most descriptive. Or `mixing_rates` with some comments around can also work to me.

Comment (akva2) [reply opm/simulators/wells/SingleWellState.hpp]: let's just make it phase_mixing_rates then.

**Comment by akva2** [comment]
jenkins build this please

**Pull request merged**